tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c

```diff
@@ -0,0 +1,457 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <arch/cpu.h>
+#include <arch/arm/cortex_m/cmsis.h>
+#include <kernel_structs.h>
+#include <offsets_short_arch.h>
+
+
+#if !defined(__GNUC__)
+#error __FILE__ goes only with Cortex-M GCC
+#endif
+
+#define PRIORITY 0
+#define BASEPRI_MODIFIED_1 0x20
+#define BASEPRI_MODIFIED_2 0x40
+#define SWAP_RETVAL        0x1234
+
+extern int __swap(unsigned int key);
+extern void z_move_thread_to_end_of_prio_q(struct k_thread *thread);
+
+static struct k_thread alt_thread;
+static K_THREAD_STACK_DEFINE(alt_thread_stack, 1024);
+
+int volatile switch_flag;
+
+struct k_thread *p_ztest_thread;
+
+_callee_saved_t ztest_thread_callee_saved_regs_container;
+
+const _callee_saved_t ztest_thread_callee_saved_regs_init = {
+	.v1 = 0x1, .v2 = 0x2, .v3 = 0x4, .v4 = 0x8,
+	.v5 = 0x10, .v6 = 0x20, .v7 = 0x40, .v8 = 0x80 };
+
+static void load_callee_saved_regs(const _callee_saved_t *regs)
+{
+	/* Load the callee-saved registers with initial arbitrary values
+	 * (from ztest_thread_callee_saved_regs_init)
+	 */
+	__asm__ volatile (
+		"mov r1, r7;\n\t"
+		"ldmia %0, {r4-r11};\n\t"
+		"mov r7, r1;\n\t"
+		:
+		: "r" (regs)
+		: "memory"
+	);
+	__DSB();
+}
+
+static void verify_callee_saved(const _callee_saved_t *src,
+		const _callee_saved_t *dst)
+{
+	/* Verify callee-saved registers are as expected */
+	zassert_true((src->v1 == dst->v1)
```

Comment:
Why not memcmp()?

Reply:
I will look into it. No particular reason :)

Reply:
Oh, now I know,   @pizi-nordic : _callee_saved struct contains _more_ than the callee-saved registers; it contains the PSP value. I  only aim at checking the save-restore integrity of v1-v8, that's why I did not trouble with memcmp and sizes, and just brute-forced it :)